### PR TITLE
Prefer expanding parenthesized expressions before operands

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary

This PR implements Black's behavior where it first splits off parenthesized expressions before splitting before operands to avoid unnecessary parentheses:

```python
# We want 
if a + [ 
	b,
	c
]: 
	pass

# Rather than
if (
    a
    + [b, c]
): 
	pass
```

This is implemented by using the new IR elements introduced in #5596. 

* We give the group wrapping the optional parentheses an ID (`parentheses_id`)
* We use `conditional_group` for the lower priority groups  (all non-parenthesized expressions) with the condition that the `parentheses_id` group breaks (we want to split before operands only if the parentheses are necessary)
* We use `fits_expanded` to wrap all other parenthesized expressions (lists, dicts, sets), to prevent that expanding e.g. a list expands the `parentheses_id` group. We gate the `fits_expand` to only apply if the `parentheses_id` group fits (because we  prefer `a\n+[b, c]` over expanding `[b, c]` if the whole expression gets parenthesized).

We limit using `fits_expanded` and `conditional_group` only to expressions that themselves are not in parentheses (checking the conditions isn't free)

## Test Plan

It increases the Jaccard index for Django from 0.915 to 0.917

## Incompatibilites

There are two incompatibilities left that I'm aware of (there may be more, I didn't go through all snapshot differences). 

### Long string literals
I  commented on the regression. The issue is that a very long string (or any content without a split point) may not fit when only breaking the right side. The formatter than inserts the optional parentheses. But this is kind of useless because the overlong string will still not fit, because there are no new split points. 

I think we should ignore this incompatibility for now

### Expressions on statement level

I don't fully understand the logic behind this yet, but black doesn't break before the operators for the following example even though the expression exceeds the configured line width

```python
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa < bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb > ccccccccccccccccccccccccccccc == ddddddddddddddddddddd
```

But it would if the expression is used inside of a condition. 

What I understand so far is that Black doesn't insert optional parentheses on the expression statement level (and a few other places) and, therefore, only breaks after opening parentheses. I propose to keep this deviation for now to avoid overlong-lines and use the compatibility report to make a decision if we should implement the same behavior.

### CODE REVIEW

**Comment by MichaReiser** [comment]
Current dependencies on/for this PR:
* main
  * **PR #5596** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/5596" target="_blank">Graphite</a> 
    * **PR #5608** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/5608" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/astral-sh/ruff/5608?utm_source=stack-comment).

File: crates/ruff_python_formatter/src/expression/mod.rs [hunk 1]
```diff
@@ -99,26 +106,56 @@ impl FormatRule<Expr, PyFormatContext<'_>> for FormatExpr {
...
+                    let parens_id = f.group_id("optional_parentheses");
+
+                    f.context_mut()
+                        .set_node_level(NodeLevel::Expression(Some(parens_id)));
 
-        f.context_mut().set_node_level(saved_level);
+                    let result = group(&format_args![
+                        if_group_breaks(&text("(")),
+                        indent_if_group_breaks(
+                            &format_args![soft_line_break(), format_expr],
```
Comment (konstin) [inline crates/ruff_python_formatter/src/expression/mod.rs:124]: why is the line break inside the indent?

File: crates/ruff_python_formatter/src/expression/mod.rs [hunk 2]
```diff
@@ -178,3 +215,238 @@ impl<'ast> IntoFormat<PyFormatContext<'ast>> for Expr {
         FormatOwnedWithRule::new(self, FormatExpr::default())
     }
 }
+
+/// Tests if it is safe to omit the optional parentheses.
+///
+/// We prefer parentheses at least in the following cases:
```
Comment (konstin) [inline crates/ruff_python_formatter/src/expression/mod.rs:221]: i think the should mention black

File: crates/ruff_python_formatter/tests/snapshots/black_compatibility@simple_cases__trailing_comma_optional_parens3.py.snap [hunk 3]
```diff
@@ -0,0 +1,77 @@
+---
+source: crates/ruff_python_formatter/tests/fixtures.rs
+input_file: crates/ruff_python_formatter/resources/test/fixtures/black/simple_cases/trailing_comma_optional_parens3.py
+---
+## Input
+
+```py
+if True:
```
[the remark this answers is not in the document]
Comment (konstin) [reply crates/ruff_python_formatter/tests/snapshots/black_compatibility@simple_cases__trailing_comma_optional_parens3.py.snap:8]: i think we do need to measure after the line break here

**Review by konstin — APPROVED** [review]

File: crates/ruff_python_formatter/tests/snapshots/black_compatibility@simple_cases__trailing_comma_optional_parens3.py.snap (hunk 3 above)
[the remark this answers is not in the document]
Comment (MichaReiser) [reply crates/ruff_python_formatter/tests/snapshots/black_compatibility@simple_cases__trailing_comma_optional_parens3.py.snap:8]: I'll leave this for now to fix later. This is a general problem when using `groups` that contain no further split points.

File: crates/ruff_python_formatter/src/expression/mod.rs (hunk 1 above)
Comment (MichaReiser) [reply crates/ruff_python_formatter/src/expression/mod.rs:124]: `Indent` is a bit counter-intuitive. Luckily, this is hidden a way by `soft_block_indent` and `block_indent`. `indent` only increments the indentation level, it doesn't insert any line breaks itself. The `soft_line_break` is necessary to then start a new line, with the new, incremented indent. 

This is why the *closing* newline is outside of the indent. We first need to decrement the indent counter, and only then start the new line.

**Comment by MichaReiser** [comment]
@MichaReiser merged this pull request with [Graphite](https://app.graphite.dev/github/pr/astral-sh/ruff/5608).

**Pull request merged**